freeipa/src/main/java/com/sequenceiq/freeipa/orchestrator/RotateSaltPasswordService.java

```diff
@@ -96,15 +111,45 @@ public void rotateSaltPassword(Stack stack) {
         }
     }
 
+    public void rotateSaltPasswordFallback(Stack stack) {
+        List<GatewayConfig> allGatewayConfig = gatewayConfigService.getNotDeletedGatewayConfigs(stack);
+        tryRemoveSaltuserFromGateways(stack, allGatewayConfig);
+
+        String newPassword = PasswordUtil.generatePassword();
+        SecurityConfig securityConfig = stack.getSecurityConfig();
+        securityConfig.getSaltSecurityConfig().setSaltPassword(newPassword);
+        try {
+            bootstrapService.reBootstrap(stack);
+            securityConfigService.changeSaltPassword(stack, newPassword);
+        } catch (CloudbreakOrchestratorException e) {
+            Set<String> gatewayConfigAddresses = allGatewayConfig.stream()
+                    .map(GatewayConfig::getPrivateAddress)
+                    .collect(Collectors.toSet());
+            LOGGER.warn("Failed to re-bootstrap gateway nodes after saltuser password delete", e);
+            String message = String.format("Failed to re-bootstrap gateway nodes after saltuser password delete. " +
+                            "Please check the salt-bootstrap service status on node(s) %s. " +
+                            "If the saltuser password was changed manually, " +
+                            "please remove the user manually with the command '%s' on node(s) %s and retry the operation.",
+                    gatewayConfigAddresses, SALTUSER_DELETE_COMMAND, gatewayConfigAddresses);
+            throw new CloudbreakServiceException(message, e);
+        }
+    }
+
+    private void tryRemoveSaltuserFromGateways(Stack stack, List<GatewayConfig> allGatewayConfig) {
+        try {
+            Set<String> targets = stack.getAllNodes().stream().map(Node::getHostname).collect(Collectors.toSet());
+            Map<String, String> response = hostOrchestrator.runCommandOnHosts(allGatewayConfig, targets, SALTUSER_DELETE_COMMAND);
+            LOGGER.debug("Saltuser delete command response: {}", response);
+        } catch (CloudbreakOrchestratorException e) {
+            LOGGER.warn("Failed to run saltuser delete command, assuming it is already deleted", e);
```

Comment:
instead of assuming wouldn't make it sense to check?

Reply:
If the current saltuser password is out of sync, we can not check this unfortunately. But when this happens the error message contains that this user should be removed manually - and if it was removed, the next step should succeed.